### DISCUSSION TITLE
vipsthumbnail: map failed (No such device)

### DISCUSSION THREAD

**Discussion by Lucas-Yuzc** [libvips/libvips#2415] [Q&A] [answered]
Using vipsthumbnaill to get  thumbnail  from GIF picture, got some error. I want to know What operating conditions will cause the following error.  Thanks.

**[error message]**
```
 (vipsthumbnail:6943): VIPS-WARNING **: 16:36:22.507: map failed (No such device), running very low on system resources, expect a crash soon
vipsthumbnail: unable to thumbnail /tmp/zfuse/18366184150ba/transcode/picture/gif/IMG_4858.GIF
vips_mapfile: unable to mmap
unix error: No such device
```
**[libvips version]**
libvips 8.11.0-Wed Jun  9 12:23:20 UTC 2021

**[source]**
![IMG_4858](https://user-images.githubusercontent.com/25133181/131099476-37d9dd1a-7878-4b53-aa80-763ac038bfcf.GIF)

**Comment by jcupitt** [comment, marked as answer]
Hi @Lucas-Yuzc,

The libvips GIF loader memory maps the input file before starting analysis. Your GIF is about 5mb, so I suppose the process that ran the thumbnail operation had less than 5mb of virtual memory left.

You can see how much virtual memory a process is using in `top` under the VIRT column. Most processors can have many, many TB of vmem, so I'm surprised you were able to run out. Perhaps there is some resource limit set somewhere? Maybe you have a  `ulimit` set?

**Reply by Lucas-Yuzc** [reply]
let me check for this. thank you.

**Reply by Lucas-Yuzc** [reply]
> Hi @Lucas-Yuzc,
> 
> The libvips GIF loader memory maps the input file before starting analysis. Your GIF is about 5mb, so I suppose the process that ran the thumbnail operation had less than 5mb of virtual memory left.
> 
> You can see how much virtual memory a process is using in `top` under the VIRT column. Most processors can have many, many TB of vmem, so I'm surprised you were able to run out. Perhaps there is some resource limit set somewhere? Maybe you have a `ulimit` set?

I found the reason is that the mergerfs I using disabled mmap function. 
any way to make vips not use mmap and still work?
thanks~

**Reply by jcupitt** [reply]
I'll have a look at adding this as a feature.

You can workaround this in the meantime by piping the GIF into vipsthumbnail, eg.:

```
$ cat 3198.gif | vipsthumbnail stdin[n=-1] -o x.gif 
```

Now it won't attempt to map the input (since you can't map pipes).

**Reply by jcupitt** [reply]
I added a thing to git master libvips to make it fall back to read if mmap fails. Would you be able to test this?

https://github.com/libvips/libvips/commit/cfd9dc3f01f2cca8f6dda6ab646982f179e17b0c

**Reply by Lucas-Yuzc** [reply]
> I'll have a look at adding this as a feature.
> 
> You can workaround this in the meantime by piping the GIF into vipsthumbnail, eg.:
> 
> ```
> $ cat 3198.gif | vipsthumbnail stdin[n=-1] -o x.gif 
> ```
> 
> Now it won't attempt to map the input (since you can't map pipes).

I'll try. Thanks for your reply
